Log

YruamaLairba opened the pull request:
I started to work on the log spec.

Philipp91 commented:
Ah this looks like the solution to #91. Could you add a usage example (perhaps in the dummy unit test)?

YruamaLairba commented:
> Ah this looks like the solution to #91. Could you add a usage example (perhaps in the dummy unit test)?

sorry, this branch is far behind the rest of the main branch. I don't remember the details, but i stopped this branch because at the moment, there is no way to implement fully an properly in pure rust. If you need it to display some debug information during plugin development, you can use "println!" instead and run the host from a console.

Philipp91 commented:
Thanks!

prokopyl commented:
Hi @YruamaLairba , do you still want to work on this PR?

While I agree full realtime formatting support is not easy (and, in fact, may very well be the job of a whole separate crate), I believe a simple function that just sends a nul-terminated string as a log to the host would be quite useful already. We can add the formatting part later. 🙂

YruamaLairba commented:
Lol, i read the [Log spec API](https://lv2plug.in/doc/html/group__log.html) again, and in general, this should not be used in real-time context (does the doc have been updated ?). The only exception is for debugging purpose (trace) where messing real-time is not a concern.
So formatting from plugin side instead of host side doesn't seems to be a real issue. I also think it's acceptable to support trace in real-time context through an optional feature.
i think i will look at it soon since only few change are required

prokopyl commented:
Yes, I agree, realtime formatting doesn't seem like a very big concern for now. ^^
I think having a simple API like the one you have implemented already that just logs a C string, and letting users in charge of formatting/acllocations should be enough, as I believe most errors would happen during instantiation or other non-realtime contexts.

prokopyl commented on log/src/lib.rs:
```diff
@@ -0,0 +1,119 @@
+    fn get(self) -> u32;
+}
+
+unsafe impl EntryType for URID<ErrorClass> {
```
I believe it's better to implement `EntryType` on each of the entry type structs (`TraceClass`, `ErrorClass`) and then receive an `URID<impl EntryType>` rather than an `impl EntryType`.

Not only this makes you have to write less code (you don't have to reimplement `get()` every time), but also the method still receives a full URID with all of its capabilities, and if anything it's more in line with what we do everywhere else that uses URIDs within the `lv2` crate (features, extensions, atom types, etc.) :slightly_smiling_face:

YruamaLairba replied:
thanks, i missed this trick.

prokopyl commented on log/src/lib.rs:
```diff
@@ -0,0 +1,119 @@
+use std::os::raw::*; //get all common c_type
+use urid::*;
+
+pub struct EntryClass;
```
I think having the entry types in a submodule (`entries` for instance) would be a cleaner, although it's not a lot of code, it'll show up as quite a few types in the rustdoc and might be a bit overwhelming for new users that might want to focus elsewhere (mainly the `Log` feature, or the entry type trait).

prokopyl commented on log/src/lib.rs:
```diff
@@ -0,0 +1,119 @@
+use urid::*;
+
+pub struct EntryClass;
+unsafe impl UriBound for EntryClass {
```
Looking at the docs again (http://lv2plug.in/ns/ext/log#Entry), it seems `Entry` is only the superclass of all entry types, and not an entry type itself. If anything, `Entry` should be what the trait is named, this type should probably not exist, and entry types can just be named `Error`, `Warning` without the `Class` suffix

YruamaLairba replied:
I used the class suffix just because some part of the LV2 spec use same name for  "Class", "properties" or "instance" even if there is no issue here.

prokopyl replied:
LV2's Classes, Properties and Instances are just RDF data for the manifest (`.ttl`) files, they have nothing to do with the types in C or Rust code. You can just rename them without the class suffix. :slightly_smiling_face: 

`Entry` (or `EntryClass` as you wrote it there) is not actually used in code, so it can be left out entirely as well!

YruamaLairba replied:
Did some cleaning on renaming

prokopyl commented on log/src/lib.rs:
```diff
@@ -0,0 +1,119 @@
+}
+
+impl<'a> Log<'a> {
+    pub fn print(&self, entry_type: impl EntryType, message: &[u8]) -> Result<(), ()> {
```
Using a `&[u8]` seems unsafe, as it does not guarantee that it is null-terminated. Using `&CStr` here would probably be better. :slightly_smiling_face: 

Also, since this method is quite limited and not the full implementation of the underlying print, it should probably be named `print_cstr`, that way we can add a full `print` method later without it being a breaking change.

YruamaLairba replied:
i changed  this, now i create a String from the &str and add the missing "\0"

prokopyl reviewed (changes requested):
I'll also do a quick review here, since you plan on continuing the work you may find this helpful :slightly_smiling_face:

YruamaLairba commented:
@prokopyl wait a minute, i just looked in again and it appear i have some unpushed work in my local dir. I will do a push force with my work on the top the develop branch for a cleaner git  history.

I also reminded a big interrogation i had about threading class of this feature. This feature is allowed in any non-realtime context, i think this implies the "worker" thread, but i really don't see how to implement that without messing the principle of separation of LV2 extension.

YruamaLairba commented:
Sorry, rebasing was a very bad idea, i didn't see the code wasn't compiling at all after therebase

YruamaLairba commented:
except documentation, i don't see what more i can do...

YruamaLairba commented:
@prokopyl have you an opinion on code i did?

prokopyl commented on log/src/lib.rs:
```diff
@@ -0,0 +1,167 @@
+
+/// Errors potentially generated by [`Log`](struct.Log.html) methods.
+#[derive(PartialEq, Eq, Clone, Copy, Debug)]
+pub enum LogError {
```
The LV2 Spec doesn't define what error types can be called yet, so this probably should be an empty struct instead:

```rust
#[derive(PartialEq, Eq, Clone, Copy, Debug)]
pub struct LogError {
    _private: ()
}
```

(the private field is here in case error codes are added in the future).

Also, because this is an error type, this must also implement the [`Error`](https://doc.rust-lang.org/std/error/trait.Error.html) trait (see the Rust API Guidelines: https://rust-lang.github.io/api-guidelines/interoperability.html#error-types-are-meaningful-and-well-behaved-c-good-err )

YruamaLairba replied:
Lol, i think you didn't understood what was the intention here. I did that because previous implementation could fail for different reason. Now, only printing error can happen, so i replaced LogError by an empty PrintError struct and I implemented the Error trait on it.

prokopyl commented on log/src/lib.rs:
```diff
@@ -0,0 +1,167 @@
+    // practice, but it's acceptable to use it for debugging purpose
+    // So, at this time, i just giving access to it instanciation class
+    unsafe fn from_feature_ptr(feature: *const c_void, class: ThreadingClass) -> Option<Self> {
+        match class {
```
To answer your question here: this feature can be used in any threading class, so the `class` parameter can just be ignored. :slightly_smiling_face:

YruamaLairba replied:
yes, the features can be used in any threading class, but only "trace" should be used in audio thread, and it's only intended for debug reason. At the moment, to handle this, i only see optional feature to disable/enable this feature in audio context

YruamaLairba replied:
lol, mistake, i confused rt-context with audio thread context. a Rt-safety paragraph should be added somewhere

prokopyl commented on log/src/lib.rs:
```diff
@@ -0,0 +1,167 @@
+}
+
+#[cfg(test)]
+mod tests {
```
There should probably be some tests in there (ideally in another file in the `tests` directory), just to make sure the API is at least functional. :slightly_smiling_face:

YruamaLairba replied:
Is it enought to check if a basic example can compile ? In this project it's frequently more difficult to setting up everything for the test than writing an actual plugin.

prokopyl commented:
> @prokopyl have you an opinion on code i did?

@YruamaLairba Yes! Sorry for the wait, I just did another review. :slightly_smiling_face: 

Also, don't bother with making extensive documentation for now, I'll merge this into a `log` branch on this repo so that it can be written later!

YruamaLairba commented on log/src/lib.rs:
```diff
@@ -0,0 +1,167 @@
+    /// operation, but the host may implement an option to display them for debugging purposes.
+    /// This entry type is special in that it may be written to in a real-time thread. It is
+    /// assumed that if debug tracing is enabled, real-time considerations are not a concern.
+    pub fn print(&self, entry_type: URID<impl EntryType>, message: &str) -> Result<(), LogError> {
```
done.

YruamaLairba commented on log/src/lib.rs:
```diff
@@ -0,0 +1,167 @@
+{
+}
+
+unsafe impl EntryType for URID<ErrorClass> {}
```
ok, fixed. I did a "test" checking the `print_cstr` interface.

YruamaLairba commented on log/src/lib.rs:
```diff
@@ -0,0 +1,167 @@
+        let printf = if let Some(printf) = self.internal.printf {
+            printf
+        } else {
+            return Err(LogError::NoCallback);
```
will try to do it later, doing this require change in internal feature representation.

YruamaLairba replied:
@prokopyl this change seems more tricky than i initially thought: 
 
 - I wanted to store only part of feature i really use, but...
 -  i realized i was assuming the feature was constant for the plugin lifetime.
 - I checked the spec about it and i found feature always appeared `const LV2_Feature *const *features`, letting me think feature was effectively constant but...
 -  I remembered a trap about the C/C++ `const` keywords, it's doesn't mean "constant", it just means "you can't change this". In other words, it can be modified elsewhere

YruamaLairba replied:
Finally done, i replaced the `lv2_sys::LV2_Log_Log` by an equivalent without the Option<>

YruamaLairba commented:
I think it's almost, it's only need documentation.

I wouldn't do submodule, there is too few type in my opinion.

YruamaLairba commented:
I completed and reworked the documentation.

I also removed the unsafety of the `Entry` trait. This is because the Log spec allow to define and use additional log level (i missed this earlier). I think this imply host should well behave when attempting to log with an unrecognized urid. I also add the URIbound as trait bound for `Entry`.

YruamaLairba commented:
@prokopyl you may didn't see, i'm ready for a review since a while.
